main.js

```diff
@@ -1,67 +1,125 @@
+    state.filteredProducts = [...state.products];
+    updateCart();
+    attachEventListeners();
+    updateProductGrid(state.products);
+}
+
+// Event Listeners
+function attachEventListeners() {
+    // Event delegation for product grid
+    elements.productGrid.addEventListener('click', e => {
+        if (e.target.matches('.add-to-cart')) {
+            handleAddToCart();
+        }
     });
-});
 
-// Update cart count
+    // Filter and sort buttons
+    document.getElementById('filterBtn').addEventListener('click', () => {
+        filterProducts(product => product.price < 100);
+    });
+
+    document.getElementById('sortBtn').addEventListener('click', () => {
+        const sortedProducts = [...state.filteredProducts].sort((a, b) => a.price - b.price);
+        updateProductGrid(sortedProducts);
+    });
+
+    document.getElementById('resetBtn').addEventListener('click', resetFilters);
+
+    // Price range filter
+    elements.maxPrice.addEventListener('input', handlePriceFilter);
+}
+
+// Cart functionality
+function handleAddToCart() {
+    state.cartCount++;
+    localStorage.setItem(CONFIG.STORAGE_KEY, state.cartCount);
+    updateCart();
+    showNotification('Item added to cart!');
+}
+
 function updateCart() {
-    document.getElementById('count').textContent = cartCount;
+    elements.cartCount.textContent = state.cartCount;
 }
 
-// Show notification
-function showNotification() {
-    const notification = document.getElementById('notification');
-    notification.style.display = 'block';
+// Notification system
+function showNotification(message) {
+    elements.notification.textContent = message;
+    elements.notification.style.display = 'block';
+    
     setTimeout(() => {
-        notification.style.display = 'none';
-    }, 2000);
+        elements.notification.style.display = 'none';
+    }, CONFIG.NOTIFICATION_DURATION);
+}
+
+// Product filtering
+function filterProducts(filterFn) {
+    state.filteredProducts = state.products.filter(filterFn);
+    updateProductGrid(state.filteredProducts);
 }
 
-// Sort products
-document.getElementById('sortBtn').addEventListener('click', () => {
-    const sortedProducts = [...products].sort((a, b) => a.price - b.price);
-    updateProductGrid(sortedProducts);
-});
+function handlePriceFilter(e) {
+    const maxPrice = parseInt(e.target.value);
+    state.maxPrice = maxPrice;
+    elements.priceValue.textContent = `$${maxPrice}`;
+    filterProducts(product => product.price <= maxPrice);
+}
 
-// Filter products
-document.getElementById('filterBtn').addEventListener('click', () => {
-    const filteredProducts = products.filter(product => product.price < 100);
-    updateProductGrid(filteredProducts);
-});
+// Reset filters
+function resetFilters() {
+    state.maxPrice = CONFIG.DEFAULT_MAX_PRICE;
+    elements.maxPrice.value = CONFIG.DEFAULT_MAX_PRICE;
+    elements.priceValue.textContent = `$${CONFIG.DEFAULT_MAX_PRICE}`;
+    state.filteredProducts = [...state.products];
```

Comment:
We might want to consider using a more immutable approach here, maybe with Object.freeze() for the base products array.